### ISSUE TITLE
Add flag for matrix-form B to `torch.linalg.solve`

### ISSUE THREAD

**Issue by wbeardall** [pytorch/pytorch#120402]
### 🚀 The feature, motivation and pitch

When `torch.linalg.solve` is presented with a B tensor of size `[n, n]` and a batched A tensor with batch size `b == n`, it assumes that B represents a batch of vectors. If batch size `b != n`, then `linalg.solve` works as expected, correctly inferring that B is, in fact, a matrix with no batch dims. This creates unexpected failures when changing batch size to `b == n`, as shown below.

When utilising `torch.linalg.solve`, it is generally known whether `B` of size `[n, n]` should be treated as a matrix or a batch of vectors. There is, however, no flag to represent this case. We would like to request a `matrix_B=True` flag in the `torch.linalg.solve` interface to allow engineers to pre-specify this behaviour.

### Alternatives

The problem can be solved by `torch.vmap` alongside Python logic for falling back to the standard `torch.linalg.solve`. However, this solution is inelegant at best, and rather cumbersome as an additional implementation consideration for developers.

A `vmap` implementation is given below, and an accuracy test at the bottom of "Additional Context"

```
def linalg_solve_with_flag(A, B, *, left: bool = True, matrix_B: bool = False):
    if matrix_B and B.ndim == 2:
        return torch.vmap(lambda A: torch.linalg.solve(A, B, left=left))(A)
    return torch.linalg.solve(A, B, left=left)
```

### Additional context

We provide a demonstration of failure modes for illustrative purposes.

```
from itertools import product

import torch

def test_linalg_solve(b, n, left):
    A = torch.rand(b, n, n)
    B = torch.rand(n, n)
    return torch.linalg.solve(A, B, left=left)

opts = [16, 32, 64]

# torch.linalg.solve defaults to assuming B is a batch of vectors. There is no way to change this
# behaviour, such as a 'matrix_B=True' flag, so in cases where batch size == n, linalg.solve fails
# in one of two ways, depending on 'left':

def scan_sizes(solve_fn):
    for b, n, left in product(opts, opts, [False, True]):
        msg = f"n = {n}, batch size = {b}, left = {left}"
        expected_shape = torch.Size([b, n, n])
        try:
            out = solve_fn(b, n, left)
        except RuntimeError:
            print(f"❌ RuntimeError for {msg}")
            continue
        if out.shape != expected_shape:
            print(f"❌ Shape {tuple(out.shape)} != {tuple(expected_shape)} for {msg}")
        else:
            print("✅")
    print("\n")

print("Testing standard linalg solve...")
scan_sizes(test_linalg_solve)
```

```
Testing standard linalg solve...
❌ RuntimeError for n = 16, batch size = 16, left = False
❌ Shape (16, 16) != (16, 16, 16) for n = 16, batch size = 16, left = True
✅
✅
✅
✅
✅
✅
❌ RuntimeError for n = 32, batch size = 32, left = False
❌ Shape (32, 32) != (32, 32, 32) for n = 32, batch size = 32, left = True
✅
✅
✅
✅
✅
✅
❌ RuntimeError for n = 64, batch size = 64, left = False
❌ Shape (64, 64) != (64, 64, 64) for n = 64, batch size = 64, left = True
```

This can be solved by vmapping the linalg solve, BUT this only works if you know
a priori B will not be batched (or by using Python logic), which is not ideal:

```
def matrix_B_solve(A, B, left):
    return torch.vmap(lambda A: torch.linalg.solve(A, B, left=left))(A)

def test_linalg_vmap_solve(b, n, left):
    A = torch.rand(b, n, n)
    B = torch.rand(n, n)
    return matrix_B_solve(A, B, left=left)

print("Testing vmapped linalg solve...")
scan_sizes(test_linalg_vmap_solve)
```

```
Testing vmapped linalg solve...
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
✅
```

We reiterate the `vmap` and check its accuracy:

```
def linalg_solve_with_flag(A, B, *, left: bool = True, matrix_B: bool = False):
    if matrix_B and B.ndim == 2:
        return torch.vmap(lambda A: torch.linalg.solve(A, B, left=left))(A)
    return torch.linalg.solve(A, B, left=left)


import time

n = 128
A = torch.rand(n, n, n)
B = torch.rand(n, n)

_ = linalg_solve_with_flag(A, B)
t = time.time()
out = linalg_solve_with_flag(A, B, matrix_B=True)
print(f"Flagged version: {time.time() - t}")

for i in range(n):
    ref = torch.linalg.solve(A[i], B)

    assert torch.all(torch.isclose(out[i], ref)), f"{(out[i] - ref).abs().sum()}"

```

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano

**Comment by wbeardall** [comment]
Similarly, utilising `torch.linalg.solve(A, B.broadcast_to(A.shape), ...)` resolves this condition. I don't imagine that the extra interface complexity is worth the minor inconvenience / consistency bonus w.r.t the `b == n` case.
